ecommerce/models.py

```diff
@@ -0,0 +1,289 @@
+        return f"ProductVersion for {self.description}, ${self.price}"
+
+
+class Basket(TimestampedModel):
+    """
+    Represents a User's basket. A Basket is made up of BasketItems. Each Basket is assigned to one user and
+    it is reused for each checkout.
+    """
+
+    user = models.OneToOneField(settings.AUTH_USER_MODEL, on_delete=models.PROTECT)
+
+    def __str__(self):
+        """Description of Basket"""
+        return f"Basket for {self.user}"
+
+
+class BasketItem(TimestampedModel):
+    """
+    Represents one or more products in a user's basket.
+    """
+
+    product = models.ForeignKey(Product, on_delete=models.PROTECT)
+    basket = models.ForeignKey(
+        Basket, on_delete=models.PROTECT, related_name="basketitems"
+    )
+    quantity = models.PositiveIntegerField()
+
+    def __str__(self):
+        """Description of BasketItem"""
+        return f"BasketItem of product {self.product} (qty: {self.quantity})"
+
+
+class Order(TimestampedModel, AuditableModel):
+    """
+    An order containing information for a purchase. Orders which are fulfilled represent successful
+    completion of a purchase and are the source of truth for this information.
+    """
+
+    FULFILLED = "fulfilled"
+    FAILED = "failed"
+    CREATED = "created"
+    REFUNDED = "refunded"
+
+    STATUSES = [CREATED, FULFILLED, FAILED, REFUNDED]
+    FULFILLED_STATUSES = [FULFILLED]
+
+    purchaser = models.ForeignKey(
+        settings.AUTH_USER_MODEL, on_delete=models.PROTECT, related_name="orders"
+    )
+    status = models.CharField(
+        choices=[(status, status) for status in STATUSES],
+        default=CREATED,
+        max_length=30,
+        db_index=True,
+    )
+
+    def __str__(self):
+        """Description for Order"""
+        return f"Order #{self.id}, status={self.status}"
+
+    @classmethod
+    def get_audit_class(cls):
+        return OrderAudit
+
+    def to_dict(self):
+        """
+        Get a serialized representation of the Order and any attached Basket and Lines
+        """
+        raise NotImplementedError
+
+
+class OrderAudit(AuditModel):
+    """
+    Audit model for Order. This table is only meant for recordkeeping purposes. The serialized
+    orders will also include information from any related tables.
+    """
+
+    order = models.ForeignKey(Order, null=True, on_delete=models.PROTECT)
+
+    @classmethod
+    def get_related_field_name(cls):
+        return "order"
+
+
+class Line(TimestampedModel):
+    """
+    A line in an Order. This contains information about a specific item which is purchased.
+    """
+
+    order = models.ForeignKey(Order, on_delete=models.PROTECT, related_name="lines")
+    product_version = models.ForeignKey(ProductVersion, on_delete=models.PROTECT)
+    quantity = models.PositiveIntegerField()
+
+    def __str__(self):
+        """Description for Line"""
+        return f"Line for order #{self.order.id}, {self.quantity} {self.product_version}(s)"
```

Comment:
Same deal with the other comment about `__str__` implementation with quantity

Reply:
:+1: